### PR TITLE
Issue #3607: Use reflection to load Checks base on checkstyle_package…

### PR DESCRIPTION
**Pull request by Luolc**
#3607 

At first I didn't introduced reflections lib, instead used guava to get the classes. But later I saw #3455 , then I tried reflections. But in fact guava is also a dependency of reflections. :(

To be honest, I don't think the current fix is good enough. Please give me any suggestions that could avoid using reflections lib. 

Do I need to generate a diff report for this fix?

In addition, I am not sure whether the CI would pass or not. `mvn clean verify` success on my local. But the behavior of reflections seems a bit strange, sometimes.

### CODE REVIEW

**Comment by MEZk** [comment]
@Luolc 
https://github.com/checkstyle/checkstyle/issues/3607 is under discussion.

File: pom.xml [hunk 1]
```diff
@@ -252,14 +252,19 @@
     <dependency>
       <groupId>com.google.guava</groupId>
       <artifactId>guava</artifactId>
-      <version>19.0</version>
+      <version>20.0</version>
```
Comment (MEZk) [inline pom.xml]: you cannot update the version just to satisfy reflections. It should be a reasonable derision. Each version update should be a separate commit.

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java [hunk 2]
```diff
@@ -239,463 +245,66 @@ private Object createObject(String className) throws CheckstyleException {
     }
 
     /**
-     * Fill short-to-full module names map.
-     */
-    private static void fillShortToFullModuleNamesMap() {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: I don't believe we are expecting these hardcoded lists to be removed, but we are still waiting more word from @romani .

**Comment by Luolc** [comment]
@MEZk @rnveach 
If the issue is no longer approved then, I am ok to abort this PR. 😄  I just had some work before the discussion and think it might be help if I share the current implementation to public.

File: pom.xml (hunk 1 above)
Comment (romani) [inline pom.xml]: there is problem with guava 20 upgrade , see #3598 .

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java (hunk 2 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: hardcoded list should stay to avoid bruteforce search, is is a problem to keep it ?

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java [hunk 3]
```diff
@@ -239,463 +245,66 @@ private Object createObject(String className) throws CheckstyleException {
...
     /**
-     * Fill short-to-full module names map with modules from filefilters package.
+     * Whether the class should be added to the map.
+     * @param clazz the class to check
+     * @return true if the class should be added to the map
      */
-    private static void fillModulesFromFilefiltersPackage() {
-        NAME_TO_FULL_MODULE_NAME.put("BeforeExecutionExclusionFileFilter",
-                BASE_PACKAGE + ".filefilters.BeforeExecutionExclusionFileFilter");
+    private static boolean shouldAddToMap(Class<?> clazz) {
```
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: @romani @Luolc 
Personally, I don't like this approach. We add so many conditions and new logic just for one thing. What are the benefits? We do not introduce new checks very frequently, so we can simply add the new checks into the map.

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java [hunk 4]
```diff
@@ -239,463 +245,66 @@ private Object createObject(String className) throws CheckstyleException {
...
-                BASE_PACKAGE + ".filters.SeverityMatchFilter");
-        NAME_TO_FULL_MODULE_NAME.put("SuppressionCommentFilter",
-                BASE_PACKAGE + ".filters.SuppressionCommentFilter");
-        NAME_TO_FULL_MODULE_NAME.put("SuppressionFilter",
-                BASE_PACKAGE + ".filters.SuppressionFilter");
-        NAME_TO_FULL_MODULE_NAME.put("SuppressWarningsFilter",
-                BASE_PACKAGE + ".filters.SuppressWarningsFilter");
-        NAME_TO_FULL_MODULE_NAME.put("SuppressWithNearbyCommentFilter",
-                BASE_PACKAGE + ".filters.SuppressWithNearbyCommentFilter");
+    private static boolean isNameCorrespondingToCorrectPackage(
```
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: Once again, new conditions will result in bugs in future.

**Comment by rnveach** [comment]
@romani Until specified otherwise, we are still trying to complete this PR?

**Comment by romani** [comment]
@Luolc , please keep this branch and please propose solution withtout extra dependency, new PR, if our code become ugly to do scan what we need, I still consider to use thirparty library (as last resort only).

**Comment by Luolc** [comment]
@romani Sorry for my absence these days. My hand injured a couple of days ago and not able to code a lot recently... It might take a week to recover.

**Comment by rnveach** [comment]
Closing this as we merged Guava's version
